### PR TITLE
react-components:Add log management UI

### PR DESCRIPTION
**Pull request by matiasbavera**
### React Components
* Added log table component customize to show different colors depending on the Log Level.
* Added a log search form.
* * Added a log management component.
* Added stories for the `SearchForm`, the `LogTable`, and `LogManagement` component.

image

## Self-checks

- [X] I'm familiar with and follow this [ Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [X] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion
* I added a pagination component to handle the logs, but if the pagination it's counter-intuitive we can change it to an infinite-scroll when we handle the pagination in the backend.

### CODE REVIEW

**Comment by koonpeng** [comment]
https://kubernetes.io/docs/concepts/cluster-administration/logging/ This should give us some idea of the logging architecture we should aim for.

**Comment by matiasbavera** [comment]
image
image

**Comment by koonpeng** [comment]
this looks great, we can merge this in react components first while we develop the backend to make it useful in the dashboard

**Comment by matiasbavera** [comment]
Adding some tests now. I'll move the fluent conf to another PR

**Comment by matiasbavera** [comment]
Hi @koonpeng , ready for the first iteration =)

File: packages/react-components/stories/log-table.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { Story, Meta } from '@storybook/react';
```
Comment (koonpeng) [inline packages/react-components/stories/log-table.stories.tsx:1]: image

image

The column size is changing across pages, there is also too much space given for "Level" sometimes. Having a long message in one row also causes all the rows to take more spacing, this causes some side effects like the date being split across multiple lines.

File: packages/react-components/stories/search-log-form.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { Story, Meta } from '@storybook/react';
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/react-components/stories/search-log-form.stories.tsx:1]: Should we replace/combine this with the log management story in dashboard?

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/react-components/lib/date-time-picker.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/lib/date-time-picker.tsx:1]: Yeah, I thought a native solution will be better for the browser support, but I guess I was wrong. =0

[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/lib/date-time-picker.tsx:1]: Had this error with Moment lib, fixed by downgrading `date-io` to 1.3.13 

```
utils.getYearText is not a function
TypeError: utils.getYearText is not a function
    at DateTimePickerToolbar (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:269584:18)
    at renderWithHooks (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:220318:18)
    at mountIndeterminateComponent (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:222997:13)
    at beginWork (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:224111:16)
    at HTMLUnknownElement.callCallback (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:205703:14)
    at Object.invokeGuardedCallbackDev (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:205752:16)
    at invokeGuardedCallback (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:205807:31)
    at beginWork$1 (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:228718:7)
    at performUnitOfWork (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:227672:12)
    at workLoopSync (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:227645:22)
```

It's a known bug. 
https://github.com/mui-org/material-ui-pickers/issues/1442


Having this warning on running test:
`
 Deprecation warning: value provided is not in a recognized RFC2822 or ISO format. moment construction falls back to js Date(), which is not reliable across all browsers and versions. Non RFC2822/ISO date formats are discouraged. Please refer to http://momentjs.com/guides/#/warnings/js-date/ for more info.`

This is on purpose because I'm removing the timezone to make those tests.

File: packages/dashboard/src/components/log-management.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/src/components/log-management.tsx:1]: Sure. I'll make sure this component can receive the callbacks from its parent then.

[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/src/components/log-management.tsx]: Done

File: packages/react-components/lib/log-management/log-table.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+}
+const useStyles = makeStyles((theme) => ({
+  table: {
+    minWidth: 650,
+  },
+  error: {
+    color: theme.palette.error.main,
+  },
+  debug: {
+    color: theme.palette.success.main,
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/lib/log-management/log-table.tsx:21]: Well, I thought that generally, the `bug` icons are green, so I placed green as a `debug` color, but I can switch it to theme.palette.secondary.dark, or to a darker green. Let me know what you think.

File: packages/react-components/lib/log-management/search-log-form.tsx [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    if (!toLogDate) {
+      setToLogDateError('Cannot be empty');
+      isValid = false;
+    }
+    return isValid;
+  };
+
+  const searchQuery = () => {
+    if (!isFormValid()) return;
+    console.log(searchText, logLabel, logLevel, rowsCount, toLogDate, fromLogDate);
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/lib/log-management/search-log-form.tsx:38]: Yeah, Thanks.

File: packages/react-components/stories/search-log-form.stories.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/stories/search-log-form.stories.tsx:1]: Makes, sense, thanks for the feedback.

[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/stories/search-log-form.stories.tsx]: image

File: packages/react-components/lib/tables/table-footer-pagination.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/lib/tables/table-footer-pagination.tsx:1]: Mmm.. sorry but maybe I'm missing something, I'm using the `TablePagination` from material UI https://next.material-ui.com/components/pagination/#table-pagination . I just wrapped it into a custom component for two things: 
 * In case we have another table that needs pagination.
 * In case we need to change the paginations of all tables later (because as you said it's in `lab` version).

If none of those reasons convince you, I can put it without problem directly in `log-table.tsx` =)

File: packages/react-components/stories/log-table.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    message: `npm ERR! code ELIFECYCLE
+    npm ERR! errno 1
+    npm ERR! react-components@0.0.1 build: 'npm run lint && tsc --build'
+    npm ERR! Exit status 1
+    npm ERR! 
+    npm ERR! Failed at the react-components@0.0.1 build script.
+    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
+    
+    npm ERR! A complete log of this run can be found in:
+    npm ERR!     /home/ekumen/.npm/_logs/2021-01-18T21_20_07_480Z-debug.log`,
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/stories/log-table.stories.tsx]: Actually, I don't know. But now I "fixed" that. But we can go back to that later if it's better.
same column width

File: packages/dashboard/src/stories/log-management.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/src/stories/log-management.stories.tsx]: sticky

File: packages/react-components/stories/log-table.stories.tsx (hunk 1 above)
Comment (matiasbavera) [reply packages/react-components/stories/log-table.stories.tsx]: Fixed. 
same width

**Comment by matiasbavera** [comment]
Rebased

**Comment by koonpeng** [comment]
Can you add some logs to the log management story so we can test the search/filter function? For the search component, do you see any use besides using it with the log management? If not then I suppose we don't really need a story for it as the log management story already covers it.

nit:
The possible of the columns sometimes change slightly when changing pages

https://user-images.githubusercontent.com/45189696/106224706-c646c600-621e-11eb-8669-0e2b4d30495e.mp4

File: packages/react-components/lib/date-time-picker.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/react-components/lib/date-time-picker.tsx:1]: >  This is on purpose because I'm removing the timezone to make those tests.

Can' we test with timezones? Is there a way to tell jsdom what locale settings to use?

File: packages/react-components/package.json [hunk 10]
```diff
@@ -17,8 +17,11 @@
   "author": "koonpeng@openrobotics.org",
   "license": "Apache-2.0",
   "dependencies": {
+    "@date-io/moment": "^1.3.13",
```
Comment (koonpeng) [inline packages/react-components/package.json:20]: Which version was it that caused the problem? npm may install newer versions based on semver.

File: packages/react-components/tests/log-management/log-management.test.tsx [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+const getLogsPromise = async () => await getLogs();
+const getLabelsPromise = async () => await getLogLabels();
+
+it('smoke test', async () => {
+  // Added the waitFor because this component is updating a state inside a useEffect.
+  await waitFor(() => {
+    render(<LogManagement getLabels={getLabelsPromise} getLogs={getLogsPromise} />);
+  });
```
Comment (koonpeng) [inline packages/react-components/tests/log-management/log-management.test.tsx:29]: Would it be better to use `act`?

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/react-components/lib/date-time-picker.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/react-components/lib/date-time-picker.tsx:1]: Fixed moment warning by moving the date to ISO format.

File: packages/dashboard/src/stories/log-management.stories.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/src/stories/log-management.stories.tsx]: Resolving this

**Comment by matiasbavera** [comment]
> Can you add some logs to the log management story so we can test the search/filter function? For the search component, do you see any use besides using it with the log management? If not then I suppose we don't really need a story for it as the log management story already covers it.

Oh sorry, now I get it, I thought that the function of that word search engine was just to send it with the request and that the backend filter the data based on that.

As an option for what I think you are saying we could implement this: https://material-ui.com/components/data-grid/
To search within the data loaded in the UI

Let me know what you think

**Comment by matiasbavera** [comment]
> nit:
> The possible of the columns sometimes change slightly when changing pages
> 
>  Peek.2021-01-29.10-41.mp4

Thanks for the feedback, fixed, the label and the content are not moving horizontally anymore. That movement you see is vertical because of the vertical space that the following message use

```
  npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! react-components@0.0.1 build: 'npm run lint && tsc --build'
    npm ERR! Exit status 1
    npm ERR! 
    npm ERR! Failed at the react-components@0.0.1 build script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/ekumen/.npm/_logs/2021-01-18T21_20_07_480Z-debug.log 
```

fixed table


Also, I'm using rem now as the unit metric, for the table to scale correctly on zoom.

File: packages/react-components/tests/log-management/log-management.test.tsx (hunk 11 above)
Comment (matiasbavera) [reply packages/react-components/tests/log-management/log-management.test.tsx:29]: Actually, from what I [read](https://github.com/testing-library/react-testing-library/issues/535#issuecomment-660267339), the RTL render function is already wrapped in an act by default. In this case, we need the waitFor because we need to wait until the `useEffect` async content is executed. 

https://testing-library.com/docs/dom-testing-library/api-async/#waitfor

Offtopic: during my search about the problem I found this interesting post https://kentcdodds.com/blog/common-mistakes-with-react-testing-library =)

File: packages/react-components/package.json (hunk 10 above)
Comment (matiasbavera) [reply packages/react-components/package.json]: Well, that's weird my npm installed 2.9.2 as the last version =O. I'll not trust npm blindly anymore. We have ^2.10.6 now, that it's the current last version. It's working correctly now after some date format config

File: packages/react-components/tests/log-management/log-management.test.tsx (hunk 11 above)
Comment (koonpeng) [reply packages/react-components/tests/log-management/log-management.test.tsx:29]: thanks for the explaination

**Comment by koonpeng** [comment]
> > Can you add some logs to the log management story so we can test the search/filter function? For the search component, do you see any use besides using it with the log management? If not then I suppose we don't really need a story for it as the log management story already covers it.
> 
> Oh sorry, now I get it, I thought that the function of that word search engine was just to send it with the request and that the backend filter the data based on that.
> 
> As an option for what I think you are saying we could implement this: https://material-ui.com/components/data-grid/
> To search within the data loaded in the UI
> 
> Let me know what you think

I haven't thought about how the searching should work but I guess a mix of server queries and client filtering would be best. We should at least be able to query the logs based on time so that we don't download the entire history. More specific filters or sorting can be done on the client to be more responsive.

The data grid looks really nice, is it fully client side or does it support dynamicly querying the server?

**Review by koonpeng — COMMENTED** [review]
I did not think about how the search form works, I instinctly thought the filters are done client side, which way do you think is better? Fully server side or a mix of server and client filtering?

Suppose we go with fully server side, then I guess it makes sense to separate the form and the viewer. I think it would still be good to have a mock data source so we can easily test out the search & filter capabilities, but that is probably a problem we need to deal with later when we do the implementation in `dashboard`.

In that case, I suppose there is no use for the log management story as it isn't useful without a backend, it makes more sense to move it back into `dashboard`. But if we do mixed server/client search & filter, then it makes sense for log management to be in react-components, it will take in an array of logs instead of a query function and we can test it in a story.

**Comment by matiasbavera** [comment]
> I did not think about how the search form works, I instinctly thought the filters are done client side, which way do you think is better? Fully server side or a mix of server and client filtering?

Yeah, I think a mix should be good. So, as you said, the main idea is to handle the heavy query in the backend, e.g. search between dates, labels (optional), and log type(optional) (dashboard, rmf-core), etc. The two things that drive me to do this is that:
* Having these pre-filters we will reduce the amount of data that we send to the front-end
* By handling less data, it will be easier for the browser to render and handle the logs in memory. 

So, when we have the response we can filter it in the UI, to make things easier for the final user.


I decided to drop this option for all the limitations that the community version has (before it did not have so many limitations) 
https://material-ui.com/components/data-grid/getting-started/#feature-comparison


And I choose https://material-table.com/#/docs/get-started which is pretty cool and configurable.

table


So I think the flow will be something like this. The search button will request data to the backend, and the table will show the response and it'll provide tools to filter and sort the data:
flow


Let me know what you think.

File: packages/react-components/package.json (hunk 10 above)
Comment (matiasbavera) [reply packages/react-components/package.json]: It's not working again, it explodes in some situations. I'll rollback to ^1.3.13. The creators of material-ui-picker also recommend using 1.x.x.
https://material-ui-pickers.dev/getting-started/installation#peer-library

**Comment by koonpeng** [comment]
I think the idea is right, some nits about the design

1. It would be nice if the log level filter has a selectable list.
2. It's not obvious to me how the time filter works, does it return results close to the time, before, or after it?
3. It's not obvious what the "search" field in the table is searching for. Is it a combination of log and message filter?
4. It can be confusing to have 2 "search" field/button. Some of which like log level and message text is duplicated, in general I think its more clear if we remove some of the existing server query params. We can also try renaming the server search button to something like "retrieve" to avoid confusion.

**Comment by matiasbavera** [comment]
> I think the idea is right, some nits about the design

Awesome!

> 1. It would be nice if the log level filter has a selectable list.

Yeah, makes sense. Done!

> 2. It's not obvious to me how the time filter works, does it return results close to the time, before, or after it?

It's the exact date, I'll find a workaround to filter between datetimes. 

> 3. It's not obvious what the "search" field in the table is searching for. Is it a combination of log and message filter?
> 4. It can be confusing to have 2 "search" field/button. Some of which like log level and message text is duplicated, in general I think its more clear if we remove some of the existing server query params. We can also try renaming the server search button to something like "retrieve" to avoid confusion.

Yeah, you're right, to avoid confusion:
* Changed `Search` -> `Retrieve Logs`
* Removed the search input component
* Removed rows count

**Comment by koonpeng** [comment]
> It's the exact date, I'll find a workaround to filter between datetimes.

How about leaving the date filtering to the server side query? By default it will retrieve the last X days, then the user can use the "retrieve logs" form to change the date range. We do lose the function of near instant time filtering but I think it is a good compromise, maybe if we really want it we can do a "smart query" where we only retrieve the logs which have not been already retrieved, if the date range falls into a previously retrieved set then we just reuse it. But that looks like a non-trivial amount of logic, we have to take care of edge cases like "refreshing" the latest logs.

**Comment by matiasbavera** [comment]
> How about leaving the date filtering to the server side query? By default it will retrieve the last X days, then the user can use the "retrieve logs" form to change the date range. We do lose the function of near instant time filtering but I think it is a good compromise, maybe if we really want it we can do a "smart query" where we only retrieve the logs which have not been already retrieved, if the date range falls into a previously retrieved set then we just reuse it. But that looks like a non-trivial amount of logic, we have to take care of edge cases like "refreshing" the latest logs.

Yeah, makes sense. We can do that. But I'd change the word `days` per `hours` because of the number of logs. Ok, I'll remove the date filter from the table and we can manage the default load from the component in the dashboard, which it'll be created when we have the logs (currently working on a PR with fluentD, I'm getting logs from containers, log files and HTTP protocol and exporting to plain text files ) and backend for it.

**Review by koonpeng — APPROVED** [review]
thanks for this, just a couple of nits

1. did the font changed?
2. We can reduce the width of the "Log Level" column.
3. Sort the log levels according to the severity
image
4. Show all the log levels without scrolling, reducing the padding/margin between the lines might help.
5. ~Would it be more intuitive if all the checkboxes are checked by default? (I think it shows all logs by default)~. It seems like it is the default behavior for MaterialTable, I guess this is more inline with how office apps works.

**Comment by matiasbavera** [comment]
> thanks for this, just a couple of nits
> 
> 1. did the font changed?
> 2. We can reduce the width of the "Log Level" column.
> 3. Sort the log levels according to the severity
>    image
> 4. Show all the log levels without scrolling, reducing the padding/margin between the lines might help.
> 5. ~Would it be more intuitive if all the checkboxes are checked by default? (I think it shows all logs by default)~. It seems like it is the default behavior for MaterialTable, I guess this is more inline with how office apps works.

1. Yeah,  you're right, good catch, that's weird because it's built on top of material-ui. But I unified the font now. 
2. Done!
image
3. Done!
4.
